demoapp/extract_mgl_sections.py

```diff
@@ -0,0 +1,319 @@
+    
+    Parameters:
+    - df (pd.DataFrame): The Pandas DataFrame containing the data.
+    - chap_sec_lists_col_name (str): The name of the column in the DataFrame that contains lists of chapter-section pairs.
+    - mgl_col_name (str): The name of the column in the DataFrame that has the text from combined MGL sections
+
+    Returns:
+    - pd.DataFrame: The modified DataFrame with an additional column 'MGL Ref' containing formatted text data
+      for each chapter-section pair in the specified column.
+    """
+    # Initialize the new columns with default values
+    df['MGL Sections List'] = None
+    df["Combined_MGL"] = "None"
+    df["No MGL Found"] = None
+    total_rows = len(df)
+    
+    # Iterate over the DataFrame rows using iterrows()
+    for index, row in df.iterrows():
+        # Calculate and print the percentage progress
+        progress = (index + 1) / total_rows * 100
+        print(f"\rProgress: {progress:.2f}%", end='')
+
+        # Get MGL sections from the API as well as chapter-section pairs with empty API call
+        mgl_list, empty_responses = get_mgl_from_api(row[chap_sec_lists_col_name])
+        df.at[index, 'MGL Sections List'] = mgl_list
+        df.at[index, 'No MGL Found'] = empty_responses
+        
+        # combine the MGL sections for one bill into one string
+        df.at[index, "Combined_MGL"] = ' '.join(mgl_list) if len(mgl_list) != 0 else "None"
+
+    return df
+    
+
+def word_count(text):
+    """
+    Calculates the number of words in the provided text.
+
+    Parameters:
+    - text (str): The input text for which the word count is to be calculated.
+
+    Returns:
+    - int: The number of words in the text. Returns 0 if the input is NaN.
+    """
+    
+    if pd.notna(text):
+        words = text.split()  # split by spaces
+        return len(words)
+    else:
+        return 0
+
+
+def get_mgl_word_count(df, word_count_col_name):
+    """
+    Applies the word_count function to a specified column in a DataFrame, creating a new column 'NumWords'.
+
+    Parameters:
+    - df (pd.DataFrame): The Pandas DataFrame containing the data.
+    - combined_mgl_col_name (str): The name of the column in the DataFrame that contains combined MGL references.
+
+    Returns:
+    - df: pd.DataFrame: The modified DataFrame with an additional column 'NumWords' containing the number of words
+      for each MGL reference in the specified column.
+
+    Note:
+    - This function utilizes the get_mgl_length function to calculate the number of words in each MGL reference
+      in the specified column and creates a new column 'NumWords' to store the results.
+    """
+    df[word_count_col_name] = df["Combined_MGL"].apply(word_count) # Column name hardcoded for now as it will be used in the demo_app.py
+    return df
+
+def get_df_with_mgl(bills_csv_file_path, bill_title_col_name, bill_num_col_name, docket_num_col_name, bill_text_col_name, chap_sec_lists_col_name, word_count_col_name):
```

Comment:
I ran out of steam to suggest the last few function signature typehints, but you get the idea :)